test/hotspot/jtreg/runtime/os/TestExplicitPageAllocation.java

```diff
@@ -0,0 +1,173 @@
+        }
+        return false;
+    }
+
+    public static int readPageCount(String filename, String pageSize) throws Exception {
+        BufferedReader d = new BufferedReader(new FileReader(filename));
+        int pageCount = Integer.parseInt(d.readLine());
+        if (pageCount == 0) {
+            System.out.println("No pages configured for " + pageSize + "kB");
+        }
+        return pageCount;
+    }
+
+    public static void doSetup() throws Exception {
+        // Large page sizes
+        File[] directories = new File(DIR_HUGE_PAGES).listFiles(File::isDirectory);
+        for (File dir : directories) {
+            String pageSizeFileName = dir.getName();
+            Matcher matcher = HUGEPAGE_PATTERN.matcher(pageSizeFileName);
+            matcher.find();
+            String pageSize = matcher.group(1);
+            assert pageSize != null;
+            int availablePages = readPageCount(DIR_HUGE_PAGES + pageSizeFileName + "/free_hugepages", pageSize);
+            int index = Long.numberOfTrailingZeros(Long.parseLong(pageSize) * 1024);
+            pageSizes[index] = true;
+            pageCount[index] = availablePages;
+        }
+        // OS vm page size
+        vmPageSizeIndex = Long.numberOfTrailingZeros(wb.getVMPageSize());
+        pageSizes[vmPageSizeIndex] = true;
+        pageCount[vmPageSizeIndex] = Integer.MAX_VALUE;
+    }
+
+    public static void testCase(int index) throws Exception {
+        String pageSize = sizeFromIndex(index);
+        ProcessBuilder pb = ProcessTools.createJavaProcessBuilder("-Xlog:pagesize",
+                                                                  "-XX:LargePageSizeInBytes=" + pageSize,
+                                                                  "-XX:+UseParallelGC",
+                                                                  "-XX:+UseLargePages",
+                                                                  "-Xmx2g",
+                                                                  "-Xms1g",
+                                                                  "-version");
+        OutputAnalyzer output = new OutputAnalyzer(pb.start());
+        output.shouldHaveExitValue(0);
+        for (int i = index; i >= vmPageSizeIndex; i--) {
+            if (pageSizes[i]) {
+                String size = sizeFromIndex(i);
+                System.out.println("Checking allocation for " + size);
+                if (!checkOutput(output, size)) {
+                    if (requiredPageCount(i) > pageCount[i]) {
+                       continue;
+                    }
+                    errorMessage += "TestCase Failed for " + size + " page allocation, ";
+                } else {
+                    System.out.println("TestCase Passed for pagesize: " + pageSize + ", allocated pagesize: " + size);
+                }
+                break;
+            }
+        }
+
+        if (errorMessage != "") {
+            throw new AssertionError(errorMessage);
+        }
```

Comment:
So this now works, but I think it is pretty hard to follow the logic with the continues and breaks and I don't see any benefit from not just throwing the `AssertionError()` when a failure occurs. 

What do you think about something like this:
```
for (int i = index; i >= vmPageSizeIndex; i--) {
    if (pageSizes[i]) {
        String size = sizeFromIndex(i);
        System.out.println("Checking allocation for " + size);
        if (checkOutput(output, size)) {
            System.out.println("TestCase Passed for pagesize: " + pageSize + ", allocated pagesize: " + size);
            // Page allocation succeeded no need to check any more page sizes.
            return;
        } else {
            // Only consider this a test failure if there are enough configured
            // pages to allow this reservation to succeeded.
            if (requiredPageCount(i) <= pageCount[i]) {
                throw new AssertionError("TestCase Failed for " + size + " page allocation. " +
                                         "Required pages: " + requiredPageCount(i) + ", " +
                                         "Configured pages: " + pageCount[i]);
            }
        }
    }
}
```

I ran this through our testing and it also works.